lib/base/tlsstream.cpp

```diff
@@ -103,3 +104,45 @@ void UnbufferedAsioTlsStream::BeforeHandshake(handshake_type type)
 	}
 #endif /* SSL_CTRL_SET_TLSEXT_HOSTNAME */
 }
+
+/**
+ * Forcefully close the connection, typically (details are up to the operating system) using a TCP RST.
+ */
+void AsioTlsStream::ForceDisconnect()
+{
+	boost::system::error_code ec;
+
+	// Close the socket. In case the connection wasn't shut down cleanly by GracefulDisconnect(), the operating system
+	// will typically terminate the connection with a TCP RST. Otherwise, this just releases the file descriptor.
+	lowest_layer().close(ec);
```

Comment:
Please tell me you're 10 billion percent sure* that any subsequent op on *this will fail gracefully, across Boost versions. I.e that close() sets the fd to -1 or similar, so that no subsequent op (which is likely a programming error) will accidentally use the freed fd number which was taken by another new socket fast enough.

*) If not, I'd just cancel here and let the destructor do its thing (which is the same as you do here explicitly).

Reply:
IMAO there are three scenarios:

1. Stuff is cleaned up quite quickly, and the destructor calls close() for us
2. Stuff is cleaned up with a little, negligible, delay which is still ok
3. Stuff leaks or is cleaned up with an unacceptable delay

In the last scenario we already have a memory leak, so an eventual FD leak is just collateral damage. Btw, we've already shut down TCP at this point.

Reply:
> I.e that close() sets the fd to -1 or similar, so that no subsequent op (which is likely a programming error)

I don't quite understand your concerns. If you trust what the `cancel()` method does, why wouldn't you do the same for `close()`? When you close the socket, why should you be concerned about the value the file descriptor (fd) would have afterwards?

> accidentally use the freed fd number which was taken by another new socket fast enough.

Sorry, but such things are **never** going to happen. We are talking about the Boost library, which bzw. Icinga isn't the only users of. This is just an imaginary scenario that won't ever happen. If it does, we can then consider alternative libraries :).

> In the last scenario we already have a memory leak, so an eventual FD leak is just collateral damage.

The whole point of this PR is to prevent unnecessary delays and memory leaks. You just want to ignore _for you negligible_ resource leaks just because we already have some? I'm really trying my best to understand your reasoning, but I just can't :).

> Btw, we've already shut down TCP at this point.

Even then, I don't see anything wrong with that. Why rely on things happening implicitly when you have the option to do it explicitly? Besides, this method won't be used only here but also in `JsonRpcConnection`s to forcibly terminate dead connections. How would canceling non-existing I/Os via `cancel()` be helpful in any way?

Reply:
> > I.e that close() sets the fd to -1 or similar, so that no subsequent op (which is likely a programming error)
> 
> I don't quite understand your concerns. If you trust what the `cancel()` method does, why wouldn't you do the same for `close()`? When you close the socket, why should you be concerned about the value the file descriptor (fd) would have afterwards?

This is my socket:

```c
int socket1 = socket(PF_INET, SOCK_STREAM, 0);
assert(socket1 == 3);
```

Now I close it, but in userspace it's still the number 3:

```c
close(socket1);
assert(socket1 == 3);
```

Then I open another one and its number accidentally matches the above one:

```c
int socket2 = socket(PF_INET, SOCK_STREAM, 0);
assert(socket2 == socket1);
```

Now, if any thread erroneously refers to `socket1` while another operates on `socket2`, good night.

That's why (aware) C programmers reset their fds to -1 to indicate a bad fd, both to themselves and to syscalls.

> > accidentally use the freed fd number which was taken by another new socket fast enough.
> 
> Sorry, but such things are **never** going to happen. We are talking about the Boost library, which bzw. Icinga isn't the only users of. This is just an imaginary scenario that won't ever happen. If it does, we can then consider alternative libraries :).

And I'm talking about Icinga, not any libs. People at Icinga also said "such things are **never** going to happen" in the past. And then v2.14.4 became necessary.

> > In the last scenario we already have a memory leak, so an eventual FD leak is just collateral damage.
> 
> The whole point of this PR is to prevent unnecessary delays and memory leaks. You just want to ignore _for you negligible_ resource leaks just because we already have some? I'm really trying my best to understand your reasoning, but I just can't :).

I didn't call any resource leaks negligible. Please read https://github.com/Icinga/icinga2/pull/10005#discussion_r1843592538 again.

> > Btw, we've already shut down TCP at this point.
> 
> Even then, I don't see anything wrong with that. Why rely on things happening implicitly when you have the option to do it explicitly? Besides, this method won't be used only here but also in `JsonRpcConnection`s to forcibly terminate dead connections. How would canceling non-existing I/Os via `cancel()` be helpful in any way?

How explicit do you want the code to be? How far[*](https://chatgpt.com/share/67376e42-8174-8004-9824-4abb3fb82b58) would you go?

Reply:
> This is my socket:
> 
> ```c
> int socket1 = socket(PF_INET, SOCK_STREAM, 0);
> assert(socket1 == 3);
> ```
> 
> Now I close it, but in userspace it's still the number 3:
> 
> ```c
> close(socket1);
> assert(socket1 == 3);
> ```
> 
> Then I open another one and its number accidentally matches the above one:
> 
> ```c
> int socket2 = socket(PF_INET, SOCK_STREAM, 0);
> assert(socket2 == socket1);
> ```
> 
> Now, if any thread erroneously refers to `socket1` while another operates on `socket2`, good night.
>
> That's why (aware) C programmers reset their fds to -1 to indicate a bad fd, both to themselves and to syscalls.

And that's all you would get if you would've tested this using the ASIO API instead of raw syscalls.

```bash
SOCKET FD BEFORE (lowest_layer().native_handle()): 7
lowest_layer().close()
SOCKET FID AFTER(lowest_layer().native_handle()): -1
```

I'm just saying, why are you questioning the use of Asio's `close()` while you are confident in using `cancel()`? As long as you use Asio's APIs, it could set the fd to whatever value it wants to indicate a bad descriptor, and I wouldn't be bothered by that.

> How explicit do you want the code to be? How far[*](https://chatgpt.com/share/67376e42-8174-8004-9824-4abb3fb82b58) would you go?

If the current implementation were that good, this PR wouldn't have existed and we wouldn't be talking about it now. But it's not! Linking to some chatgpt threads doesn't make it any better! So, let's end this here before we go down the wrong path and discuss this in person next week.

Reply:
> ```shell
> SOCKET FD BEFORE (lowest_layer().native_handle()): 7
> lowest_layer().close()
> SOCKET FD AFTER(lowest_layer().native_handle()): -1
> ```

This is perfect!

> I'm just saying, why are you questioning the use of Asio's `close()` while you are confident in using `cancel()`?

Because `cancel()` doesn't close(2), so it can't even forget to set anything to -1.

> As long as you use Asio's APIs, it could set the fd to whatever value it wants to indicate a bad descriptor, and I wouldn't be bothered by that.

Yes, but... actually no. 😅 At least no value above -1.